### PR TITLE
LiteX bump including picolibc.  

### PR DESCRIPTION
**Pull request by tcal-x**
Signed-off-by: Tim Callahan <tcal@google.com>

### CODE REVIEW

**Comment by tcal-x** [comment]
Current error, proj/hps_accel, `make clean bitstream`:
```
INFO:SoC:--------------------------------------------------------------------------------
make[2]: Entering directory '/media/tim/GIT/tcal-x/CFU-Playground/soc/build/digilent_arty.hps_accel/software/libc'
if [ -d "/media/tim/GIT/tcal-x/CFU-Playground/third_party/python/litex/litex/soc/software/libc/riscv" ]; then \
        cp /media/tim/GIT/tcal-x/CFU-Playground/third_party/python/litex/litex/soc/software/libc/riscv/* /media/tim/GIT/tcal-x/CFU-Playground/third_party/python/pythondata-software-picolibc/pythondata_software_picolibc/data/newlib/libc/machine/riscv/ ;\
fi
 CC       stdio.c.o
 CC       missing.c.o
meson /media/tim/GIT/tcal-x/CFU-Playground/third_party/python/pythondata-software-picolibc/pythondata_software_picolibc/data \
        -Dmultilib=false \
        -Dpicocrt=false \
        -Dthread-local-storage=false \
        -Dio-long-long=true \
        -Dformat-default=integer \
        -Dincludedir=picolibc/riscv64-unknown-elf/include \
        -Dlibdir=picolibc/riscv64-unknown-elf/lib \
        --cross-file cross.txt
 AR       missing.c.o
/media/tim/GIT/tcal-x/CFU-Playground/third_party/python/litex/litex/soc/software/libc/stdio.c:97:23: warning: implicit declaration of function 'FDEV_SETUP_STREAM' [-Wimplicit-function-declaration]
 static FILE __stdio = FDEV_SETUP_STREAM(dummy_putc, dummy_getc, NULL, _FDEV_SETUP_RW);
                       ^~~~~~~~~~~~~~~~~
/media/tim/GIT/tcal-x/CFU-Playground/third_party/python/litex/litex/soc/software/libc/stdio.c:97:71: error: '_FDEV_SETUP_RW' undeclared here (not in a function)
 static FILE __stdio = FDEV_SETUP_STREAM(dummy_putc, dummy_getc, NULL, _FDEV_SETUP_RW);
                                                                       ^~~~~~~~~~~~~~
In file included from /media/tim/GIT/tcal-x/CFU-Playground/third_party/python/litex/litex/soc/software/libc/stdio.c:17:
/media/tim/GIT/tcal-x/CFU-Playground/third_party/python/litex/litex/soc/software/libc/stdio.c:99:13: error: expected ')' before '->' token
 FILE *const stdout = &__stdio;
             ^~~~~~
/media/tim/GIT/tcal-x/CFU-Playground/third_party/python/litex/litex/soc/software/libc/stdio.c:100:13: error: expected ')' before '->' token
 FILE *const stderr = &__stdio;
             ^~~~~~
/media/tim/GIT/tcal-x/CFU-Playground/third_party/python/litex/litex/soc/software/libc/stdio.c:101:13: error: expected ')' before '->' token
 FILE *const stdin  = &__stdio;
             ^~~~~
/media/tim/GIT/tcal-x/CFU-Playground/third_party/python/litex/litex/soc/software/libc/stdio.c:97:13: warning: '__stdio' defined but not used [-Wunused-variable]
 static FILE __stdio = FDEV_SETUP_STREAM(dummy_putc, dummy_getc, NULL, _FDEV_SETUP_RW);
             ^~~~~~~
make[2]: *** [/media/tim/GIT/tcal-x/CFU-Playground/third_party/python/litex/litex/soc/software/libc/Makefile:68: stdio.c.o] Error 1
make[2]: *** Waiting for unfinished jobs....

ERROR: Neither directory contains a build file meson.build.
make[2]: *** [/media/tim/GIT/tcal-x/CFU-Playground/third_party/python/litex/litex/soc/software/libc/Makefile:51: libc.a] Error 1
make[2]: Leaving directory '/media/tim/GIT/tcal-x/CFU-Playground/soc/build/digilent_arty.hps_accel/software/libc'
Traceback (most recent call last):
  File "./common_soc.py", line 57, in <module>
    main()
  File "./common_soc.py", line 53, in main
    workflow.run()
  File "/media/tim/GIT/tcal-x/CFU-Playground/soc/board_specific_workflows/general.py", line 119, in run
    soc_builder = self.build_soc(soc)
  File "/media/tim/GIT/tcal-x/CFU-Playground/soc/board_specific_workflows/digilent_arty.py", line 59, in build_soc
    return super().build_soc(soc, **kwargs)
  File "/media/tim/GIT/tcal-x/CFU-Playground/soc/board_specific_workflows/general.py", line 96, in build_soc
    soc_builder.build(run=self.args.build, **kwargs)
  File "/media/tim/GIT/tcal-x/CFU-Playground/third_party/python/litex/litex/soc/integration/builder.py", line 302, in build
    self._generate_rom_software(compile_bios=use_bios)
  File "/media/tim/GIT/tcal-x/CFU-Playground/third_party/python/litex/litex/soc/integration/builder.py", line 247, in _generate_rom_software
    subprocess.check_call(["make", "-C", dst_dir, "-f", makefile])
  File "/media/tim/GIT/tcal-x/CFU-Playground/env/conda/envs/cfu-common/lib/python3.7/subprocess.py", line 363, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['make', '-C', '/media/tim/GIT/tcal-x/CFU-Playground/soc/build/digilent_arty.hps_accel/software/libc', '-f', '/media/tim/GIT/tcal-x/CFU-Playground/third_party/python/litex/litex/soc/software/libc/Makefile']' returned non-zero exit status 2.
make[1]: *** [/media/tim/GIT/tcal-x/CFU-Playground/soc/common_soc.mk:115: build/digilent_arty.hps_accel/gateware/digilent_arty.bit] Error 1
make[1]: Leaving directory '/media/tim/GIT/tcal-x/CFU-Playground/soc'
make: *** [../proj.mk:276: bitstream] Error 2
```

**Comment by tcal-x** [comment]
Update -- the issue above is resolved by doing a recursive submodule update on the new picolibc submodule.

New error:
```
In file included from /media/tim/GIT/tcal-x/CFU-Playground/third_party/python/pythondata-software-picolibc/pythondata_software_picolibc/data/newlib/libc/include/inttypes.h:16,
                 from /media/tim/GIT/tcal-x/CFU-Playground/third_party/python/pythondata-software-picolibc/pythondata_software_picolibc/data/newlib/libc/tinystdio/stdio.h:44,
                 from /media/tim/GIT/tcal-x/CFU-Playground/third_party/python/litex/litex/soc/software/libc/stdio.c:17:
/media/tim/GIT/tcal-x/CFU-Playground/third_party/python/pythondata-software-picolibc/pythondata_software_picolibc/data/newlib/libc/include/newlib.h:35:10: fatal error: picolibc.h: No such file or directory
 #include <picolibc.h>
          ^~~~~~~~~~~~
compilation terminated.
```

**Comment by tcal-x** [comment]
I did confirm that the version of `meson` that you get  (0.53.2) from `sudo apt install meson` on Ubuntu 20.04 **does not work** with picolibc.

The version that you get (0.59.1) with gLinux `sudo apt install meson`, or with pip3, does work .

**Comment by tcal-x** [comment]
Remaining issues:
* Need a universal solution for meson
* Need https://github.com/enjoy-digital/litex/pull/1046 to land, to fix the race issue (picolibc.h not found issue).
* Need some linker flag updates according to @kgugala

**Comment by kgugala** [comment]
@googlebot I consent.

**Comment by mkurc-ant** [comment]
@tcal-x @kgugala I built both gateware and software from this branch and verified that it works correctly on the Arty platform. The golden tests fot the TFLite HPS model pass:

```
Running Golden tests (check for expected outputs)
Testing input cat: Set 76800 bytes at 0x40103d50
......................................................
"Event","Tag","Ticks"
0,STRIDED_SLICE,400
1,CONV_2D,2421
2,CONV_2D,11073
3,CONV_2D,9171
4,MAX_POOL_2D,1083
5,STRIDED_SLICE,386
6,CONV_2D,2425
7,CONV_2D,11073
8,CONV_2D,9171
9,MAX_POOL_2D,1079
10,STRIDED_SLICE,385
11,CONV_2D,2423
12,CONV_2D,11072
13,CONV_2D,9172
14,MAX_POOL_2D,1084
15,STRIDED_SLICE,392
16,CONV_2D,2421
17,CONV_2D,11073
18,CONV_2D,9171
19,MAX_POOL_2D,1084
20,STRIDED_SLICE,390
21,CONV_2D,2421
22,CONV_2D,11073
23,CONV_2D,9171
24,MAX_POOL_2D,1082
25,STRIDED_SLICE,654
26,CONV_2D,4179
27,CONV_2D,20406
28,CONV_2D,18269
29,MAX_POOL_2D,2157
30,CONCATENATION,829
31,STRIDED_SLICE,328
32,CONV_2D,10143
33,CONV_2D,24269
34,MAX_POOL_2D,925
35,STRIDED_SLICE,328
36,CONV_2D,10138
37,CONV_2D,24270
38,MAX_POOL_2D,926
39,STRIDED_SLICE,730
40,CONV_2D,25644
41,CONV_2D,72410
42,MAX_POOL_2D,2766
43,CONCATENATION,510
44,CONV_2D,32198
45,MAX_POOL_2D,1243
46,CONV_2D,6687
47,MAX_POOL_2D,166
48,CONV_2D,430
49,RESHAPE,5
50,FULLY_CONNECTED,222
51,FULLY_CONNECTED,77
52,FULLY_CONNECTED,7
53,LOGISTIC,7
Perf counters not enabled.
   391M (   390886293) cycles total
OK
Testing input diagram: Set 76800 bytes at 0x40103d50
......................................................
"Event","Tag","Ticks"
0,STRIDED_SLICE,398
1,CONV_2D,2421
2,CONV_2D,11073
3,CONV_2D,9171
4,MAX_POOL_2D,1082
5,STRIDED_SLICE,387
6,CONV_2D,2425
7,CONV_2D,11073
8,CONV_2D,9171
9,MAX_POOL_2D,1083
10,STRIDED_SLICE,385
11,CONV_2D,2423
12,CONV_2D,11073
13,CONV_2D,9171
14,MAX_POOL_2D,1088
15,STRIDED_SLICE,393
16,CONV_2D,2422
17,CONV_2D,11073
18,CONV_2D,9171
19,MAX_POOL_2D,1086
20,STRIDED_SLICE,389
21,CONV_2D,2420
22,CONV_2D,11072
23,CONV_2D,9170
24,MAX_POOL_2D,1086
25,STRIDED_SLICE,654
26,CONV_2D,4179
27,CONV_2D,20407
28,CONV_2D,18270
29,MAX_POOL_2D,2158
30,CONCATENATION,829
31,STRIDED_SLICE,327
32,CONV_2D,10143
33,CONV_2D,24270
34,MAX_POOL_2D,924
35,STRIDED_SLICE,328
36,CONV_2D,10138
37,CONV_2D,24270
38,MAX_POOL_2D,923
39,STRIDED_SLICE,731
40,CONV_2D,25643
41,CONV_2D,72411
42,MAX_POOL_2D,2764
43,CONCATENATION,510
44,CONV_2D,32198
45,MAX_POOL_2D,1241
46,CONV_2D,6686
47,MAX_POOL_2D,167
48,CONV_2D,430
49,RESHAPE,5
50,FULLY_CONNECTED,223
51,FULLY_CONNECTED,77
52,FULLY_CONNECTED,8
53,LOGISTIC,7
Perf counters not enabled.
   391M (   390895689) cycles total
OK
Testing input zeroes: Zeroed 76800 bytes at 0x40103d50
......................................................
"Event","Tag","Ticks"
0,STRIDED_SLICE,404
1,CONV_2D,2421
2,CONV_2D,11073
3,CONV_2D,9172
4,MAX_POOL_2D,1067
5,STRIDED_SLICE,387
6,CONV_2D,2425
7,CONV_2D,11073
8,CONV_2D,9171
9,MAX_POOL_2D,1067
10,STRIDED_SLICE,385
11,CONV_2D,2423
12,CONV_2D,11073
13,CONV_2D,9171
14,MAX_POOL_2D,1071
15,STRIDED_SLICE,392
16,CONV_2D,2421
17,CONV_2D,11072
18,CONV_2D,9171
19,MAX_POOL_2D,1068
20,STRIDED_SLICE,389
21,CONV_2D,2420
22,CONV_2D,11072
23,CONV_2D,9170
24,MAX_POOL_2D,1068
25,STRIDED_SLICE,654
26,CONV_2D,4178
27,CONV_2D,20407
28,CONV_2D,18269
29,MAX_POOL_2D,2125
30,CONCATENATION,829
31,STRIDED_SLICE,328
32,CONV_2D,10143
33,CONV_2D,24270
34,MAX_POOL_2D,920
35,STRIDED_SLICE,328
36,CONV_2D,10138
37,CONV_2D,24270
38,MAX_POOL_2D,921
39,STRIDED_SLICE,731
40,CONV_2D,25643
41,CONV_2D,72411
42,MAX_POOL_2D,2756
43,CONCATENATION,509
44,CONV_2D,32198
45,MAX_POOL_2D,1240
46,CONV_2D,6687
47,MAX_POOL_2D,165
48,CONV_2D,430
49,RESHAPE,5
50,FULLY_CONNECTED,222
51,FULLY_CONNECTED,77
52,FULLY_CONNECTED,8
53,LOGISTIC,7
Perf counters not enabled.
   391M (   390765447) cycles total
OK
OK   Golden tests passed
---
```

File: common/Makefile [hunk 1]
```diff
@@ -46,6 +47,7 @@ DEFINE_FLAGS := $(DEFINES:%=-D %)
 SHARED_FLAGS := \
 	-march=rv32im  \
 	-mabi=ilp32 \
+	-fno-math-errno \
```
Comment (kgugala) [inline common/Makefile:50]: we should remove this - it got here by accident

File: conf/requirements.txt [hunk 2]
```diff
@@ -3,3 +3,4 @@ termcolor
 tqdm
 yapf
 requests
+meson
```
Comment (kgugala) [inline conf/requirements.txt:6]: why do we need meson here? we have it in third_party/python

File: common/Makefile (hunk 1 above)
Comment (tcal-x) [reply common/Makefile]: Removed.

File: conf/requirements.txt (hunk 2 above)
Comment (tcal-x) [reply conf/requirements.txt]: Removed.

**Review by kgugala — APPROVED** [review]
LGTM

**Pull request merged**